### DISCUSSION TITLE
LVGL support

### DISCUSSION THREAD

**Discussion by maxbrito500** [toitlang/toit#1447] [Ideas]
Hello,

I'm evaluating Toit and so far it is looking good on all factors to consider. However, I'm unable to find info about the support for GUI applications. More specifically, to run LVGL dialog windows: https://github.com/lvgl/lvgl

Is there a way to include C libraries so that we can make usage of these libraries?

Thanks.

**Comment by floitsch** [comment]
Toit has a display library/package (https://docs.toit.io/language/sdk/display), but it's not as sophisticated as lvgl.

There is a way to interact with C code, but not as directly as in some other languages. Basically, you would establish a "service" on the C side which can then be invoked using RPC. Here is our test for this: https://github.com/toitlang/toit/blob/d5122215c2c4dff1f0922ad417a2a48f1ec76991/tests/ctest/external_messaging_test.cc

Depending on how lvgl does the rendering another challenge would be of how to get the widgets onto a display. If it wants direct access to the hardware, then that would make the code less portable, less secure, and not really "Toity". If, on the other hand, it has a good abstraction for painting, then it might be straight-forward to send the display updates efficiently through the RPC to the display.
